libraries/chain/finality/qc.cpp

```diff
@@ -0,0 +1,508 @@
+      if (pending_best_qc)
+         return std::optional<qc_t>{qc_t{block_num, std::move(*active_best_qc), std::move(pending_best_qc)}};
+      return {}; // no quorum on pending_policy_sig so no qc for this block
+   }
+
+   // no pending_policy_sig so only need active
+   return std::optional<qc_t>{qc_t{block_num, std::move(*active_best_qc), {}}};
+}
+
+void open_qc_t::verify_qc(const qc_t& qc, const digest_type& strong_digest, const weak_digest_t& weak_digest) const {
+   if (qc.pending_policy_sig) {
+      EOS_ASSERT(pending_finalizer_policy, invalid_qc_claim,
+                 "qc contains pending policy signature for nonexistent pending finalizer policy");
+   } else if (pending_finalizer_policy) {
+      EOS_ASSERT(false, invalid_qc_claim,
+                 "qc does not contain pending policy signature for pending finalizer policy");
+   }
+
+   qc.active_policy_sig.verify(active_finalizer_policy, strong_digest, weak_digest);
+   if (pending_finalizer_policy) {
+      qc.pending_policy_sig->verify(pending_finalizer_policy, strong_digest, weak_digest);
+   }
+}
+
+bool open_qc_t::set_received_qc(const qc_t& qc) {
+   bool active_better = active_policy_sig.set_received_qc_sig(qc.active_policy_sig);
+   bool pending_better = false;
+   if (qc.pending_policy_sig) {
+      assert(pending_policy_sig);
+      pending_better = pending_policy_sig->set_received_qc_sig(*qc.pending_policy_sig);
+   }
+   return active_better || pending_better;
+}
+
+bool open_qc_t::received_qc_is_strong() const {
+   if (!pending_policy_sig) { // consider only active
+      return active_policy_sig.received_qc_sig_is_strong();
+   }
+   return active_policy_sig.received_qc_sig_is_strong() && pending_policy_sig->received_qc_sig_is_strong();
+}
+
+vote_status open_qc_t::aggregate_vote(uint32_t connection_id, const vote_message& vote,
+                                      block_num_type block_num, std::span<const uint8_t> finalizer_digest)
+{
+   bool verified_sig = false;
+   auto verify_sig = [&]() -> vote_status {
+      if (!verified_sig && !fc::crypto::blslib::verify(vote.finalizer_key, finalizer_digest, vote.sig)) {
+         fc_wlog(vote_logger, "connection - ${c} signature from finalizer ${k}.. cannot be verified",
+                 ("c", connection_id)("k", vote.finalizer_key.to_string().substr(8,16)));
+         return vote_status::invalid_signature;
+      }
+      verified_sig = true;
+      return vote_status::success;
+   };
+
+   auto add_vote = [&](const finalizer_policy_ptr& finalizer_policy, open_qc_sig_t& open_qc_sig) -> vote_status {
+      const auto& finalizers = finalizer_policy->finalizers;
+      auto itr = std::ranges::find_if(finalizers, [&](const auto& finalizer) { return finalizer.public_key == vote.finalizer_key; });
+      vote_status s = vote_status::unknown_public_key;
+      if (itr != finalizers.end()) {
+         auto index = std::distance(finalizers.begin(), itr);
+         if (open_qc_sig.has_voted(vote.strong, index)) {
```

Comment:
We should use the `has_voted` version which does not take a `bool strong` first param. After all we should not accept both a weak and strong vote from the same finalizer.